src/runtime/server/utils/rate-limit.ts

```diff
@@ -1,76 +1,65 @@
-    // const url = urlWithParams.split('?')[0]
-
-    // configure the route settings
-    const url = routes[splitURL] ? splitURL : '/api/*'
-    const maxRequests = routes[url].maxRequests
-    const interval = routes[url].intervalSeconds * 1000
-
-    // remove any IPs & URLs that haven't been used since interval to keep object small
-    const currentTime = Date.now()
-    Object.keys(rateLimit).forEach((key: keyof RateLimit) => {
-      Object.keys(rateLimit[key]).forEach((urlKey: string) => {
-        const item = rateLimit[key][urlKey]
-        const timeSinceLastRequest = currentTime - item.firstRequestTime
-        const routeInterval = routes[urlKey].intervalSeconds * 1000
+export function getRateLimitPayload(event: H3Event) : false | { limited: boolean; limit: number; current: number; secondsUntilReset: number } {
+  const routeRules = getRouteRules(event)
+  if (!routeRules['nuxt-rate-limit']) {
+    return false
+  }
+  const { maxRequests, intervalSeconds, route } = routeRules['nuxt-rate-limit'] as RouteRateLimitOptions
+  const intervalMs = intervalSeconds * 1000
+  const ip = getIP(event)
 
-        // remove the url
-        if (timeSinceLastRequest >= routeInterval) {
-          delete rateLimit[key][urlKey]
-        }
-      })
+  // remove any IPs & URLs that haven't been used since interval to keep object small
+  const currentTime = Date.now()
+  Object.keys(rateLimit).forEach((key: keyof RateLimit) => {
+    Object.keys(rateLimit[key]).forEach((urlKey: string) => {
+      const item = rateLimit[key][urlKey]
+      const timeSinceLastRequest = currentTime - item.firstRequestTime
+      const routeInterval = intervalSeconds * 1000
 
-      // remove the ip
-      if (!Object.keys(rateLimit[key]).length) {
-        delete rateLimit[key]
+      // remove the url
+      if (timeSinceLastRequest >= routeInterval) {
+        delete rateLimit[key][urlKey]
       }
     })
 
-    // add a rate limit object, or set to existing
-    rateLimit[ip] = rateLimit[ip] || {}
-    rateLimit[ip][url] = rateLimit[ip][url] || {
-      firstRequestTime: Number(new Date()),
-      requests: 0,
+    // remove the ip
+    if (!Object.keys(rateLimit[key]).length) {
+      delete rateLimit[key]
     }
+  })
 
-    // check if the IP & URL is rate limited, return seconds until reset if it is
-    const requests = rateLimit[ip][url].requests
-    if (requests >= maxRequests) {
-      const timeSinceFirstRequest = currentTime - rateLimit[ip][url].firstRequestTime
-      const secondsUntilReset = Math.ceil((interval - timeSinceFirstRequest) / 1000)
-      return secondsUntilReset
-    }
+  // add a rate limit object, or set to existing
+  rateLimit[ip] = rateLimit[ip] || {}
+  // we index the rate limiting based on the route rule route, this allows us to rate limit a wild card entry using the same
+  // rate limit object, eg: /api/* will use the same rate limit object for /api/foo and /api/bar
+  rateLimit[ip][route] = rateLimit[ip][route] || {
```

Comment:
an important change here is that we rate limit for the route group and not the individual rate, this is more intuitive and is less prone to abuse